maven-core/src/main/java/org/apache/maven/ReactorReader.java

```diff
@@ -178,41 +178,49 @@ private File find( MavenProject project, Artifact artifact )
         }
         // Check whether an earlier Maven run might have produced an artifact that is still on disk.
         else if ( packagedArtifactFile != null && packagedArtifactFile.exists()
-                && isPackagedArtifactUpToDate( project, packagedArtifactFile ) )
+                && isPackagedArtifactUpToDate( project, packagedArtifactFile, artifact ) )
         {
             return packagedArtifactFile;
         }
         else if ( !hasBeenPackagedDuringThisSession( project ) )
         {
             // fallback to loose class files only if artifacts haven't been packaged yet
             // and only for plain old jars. Not war files, not ear files, not anything else.
+            return determineLooseDirectoryForArtifact( project, artifact );
+        }
+
+        // The fall-through indicates that the artifact cannot be found;
+        // for instance if package produced nothing or classifier problems.
+        return null;
+    }
 
-            if ( isTestArtifact( artifact ) )
+    private File determineLooseDirectoryForArtifact( final MavenProject project, final Artifact artifact )
```

Comment:
Since you are using a build output directory, why not use this in the method name? `determineBuildOutputDirectoryForArtifact`